apps/admin/backend/src/exports/csv_results.ts

```diff
@@ -21,135 +22,292 @@ import { ScannerBatch } from '../types';
-  }
+  // Single Attributes
+  // -----------------
 
-  if (groupBy.groupByPrecinct) {
-    assert(groupSpecifier.precinctId !== undefined);
-    values.push(
-      getPrecinctById(electionDefinition, groupSpecifier.precinctId).name
-    );
-    values.push(groupSpecifier.precinctId);
+  if (metadataStructure.precinct === 'single') {
+    const precinctId = assertOnlyElement(filter.precinctIds);
+    values.push(getPrecinctById(electionDefinition, precinctId).name);
+    values.push(precinctId);
   }
 
   if (
-    isPrimaryElection &&
-    (groupBy.groupByParty || groupBy.groupByBallotStyle)
+    election.type === 'primary' &&
+    (metadataStructure.party === 'single' ||
+      metadataStructure.ballotStyle === 'single')
   ) {
     const partyId = (() => {
-      if (groupBy.groupByParty) {
-        return assertDefined(groupSpecifier.partyId);
+      if (metadataStructure.party === 'single') {
+        return assertOnlyElement(filter.partyIds);
       }
 
+      const ballotStyleId = assertOnlyElement(filter.ballotStyleIds);
       return assertDefined(
-        getBallotStyleById(
-          electionDefinition,
-          assertDefined(groupSpecifier.ballotStyleId)
-        ).partyId
+        getBallotStyleById(electionDefinition, ballotStyleId).partyId
       );
     })();
 
-    values.push(assertDefined(getPartyById(electionDefinition, partyId).name));
+    values.push(getPartyById(electionDefinition, partyId).name);
     values.push(partyId);
   }
 
-  if (groupBy.groupByBallotStyle) {
-    assert(groupSpecifier.ballotStyleId !== undefined);
-    values.push(groupSpecifier.ballotStyleId);
+  if (metadataStructure.ballotStyle === 'single') {
+    values.push(assertOnlyElement(filter.ballotStyleIds));
+  }
+
+  if (metadataStructure.votingMethod === 'single') {
+    values.push(
+      Tabulation.VOTING_METHOD_LABELS[assertOnlyElement(filter.votingMethods)]
+    );
   }
 
-  if (groupBy.groupByScanner || groupBy.groupByBatch) {
+  if (
+    metadataStructure.scanner === 'single' ||
+    metadataStructure.batch === 'single'
+  ) {
     const scannerId = (() => {
-      if (groupBy.groupByScanner) {
-        return assertDefined(groupSpecifier.scannerId);
+      if (metadataStructure.scanner === 'single') {
+        return assertOnlyElement(filter.scannerIds);
       }
 
-      return assertDefined(batchLookup[assertDefined(groupSpecifier.batchId)])
+      return assertDefined(batchLookup[assertOnlyElement(filter.batchIds)])
         .scannerId;
     })();
     values.push(scannerId);
   }
 
-  if (groupBy.groupByBatch) {
-    values.push(assertDefined(groupSpecifier.batchId));
+  if (metadataStructure.batch === 'single') {
+    values.push(assertOnlyElement(filter.batchIds));
   }
 
+  // Multi Attributes
+  // -------------------
+
+  if (metadataStructure.precinct === 'multi') {
+    values.push(
+      assertDefined(filter.precinctIds)
+        .map((id) => getPrecinctById(electionDefinition, id).name)
+        .join(',')
```

Comment:
nit: since these are meant to be human-readable, should we add a space after the commas for readability? might wrap more naturally in CSV viewers/editors too

Reply:
Good point! Will do.

Reply:
[ac44edc](https://github.com/votingworks/vxsuite/pull/3938/commits/ac44edc3c943e3da96876aadfee5b89fa7df196c)